crates/viewer/re_viewer_context/src/component_fallbacks.rs

```diff
@@ -37,87 +48,241 @@ pub enum ComponentFallbackError {
+
+    /// Maps component identifier to fallback providers.
+    component_fallback_providers: IntMap<ComponentIdentifier, ComponentFallbackProviderFn>,
+
+    /// Maps component types to fallback descriptors, used if there's no matching
+    /// fallback in `exact_fallback_providers`.
+    type_fallback_providers: IntMap<ComponentType, ComponentFallbackProviderFn>,
+}
+
+impl FallbackProviderRegistry {
+    /// Registers a fallback provider function for a given component type.
     ///
-    /// Fallbacks can be based on arbitrarily complex & context sensitive heuristics.
-    fn try_provide_fallback(
-        &self,
-        ctx: &QueryContext<'_>,
+    /// The function is expected to return the correct type for the given
+    /// component type.
+    pub fn register_dyn_type_fallback_provider(
+        &mut self,
         component: ComponentType,
-    ) -> ComponentFallbackProviderResult;
+        provider: ComponentFallbackProviderFn,
+    ) {
+        if self
+            .type_fallback_providers
+            .insert(component, provider)
+            .is_some()
+        {
+            re_log::warn!(
+                "There was already a component fallback provider registered for {component}"
+            );
+        }
+    }
+
+    /// Registers a fallback provider function for a given component type.
+    pub fn register_type_fallback_provider<C: re_types::Component>(
+        &mut self,
+        f: impl Fn(&QueryContext<'_>) -> C + Send + Sync + 'static,
+    ) {
+        self.register_dyn_type_fallback_provider(
+            C::name(),
+            Box::new(move |query_context| {
+                let value = f(query_context);
+
+                C::to_arrow([Cow::Owned(value)])
+            }),
+        );
+    }
+
+    /// Registers a fallback provider function for a given component identifier.
+    pub fn register_dyn_fallback_provider(
+        &mut self,
+        identifier: ComponentIdentifier,
+        provider: ComponentFallbackProviderFn,
+    ) {
+        if self
+            .component_fallback_providers
+            .insert(identifier, provider)
+            .is_some()
+        {
+            re_log::warn!(
+                "There was already a component fallback provider registered for {identifier}"
+            );
+        }
+    }
+
+    /// Registers a fallback provider function for a given component identifier.
+    pub fn register_fallback_provider<C: re_types::Component>(
+        &mut self,
+        descriptor: &ComponentDescriptor,
+        provider: impl Fn(&QueryContext<'_>) -> C + Send + Sync + 'static,
+    ) {
+        debug_assert_eq!(descriptor.component_type, Some(C::name()));
+
+        self.register_dyn_fallback_provider(
+            descriptor.component,
+            Box::new(move |query_context| {
+                let value = provider(query_context);
+
+                C::to_arrow([Cow::Owned(value)])
+            }),
+        );
+    }
```

Comment:
naming of this one is a bit weird. I mean isn't it the exact same as `register_type_fallback_provider` only that it takes a descriptor?
I think if possible we should just kill that one off since it's confusing - we're not actually indexing on anytyhing else but the component identifier